tests/test_dataclasses.py

```diff
@@ -1439,25 +1446,100 @@ class MyDataclass:
     ]
 
 
-@pytest.mark.xfail(reason='need to make it possible to rebuild dataclasses')
-def test_cyclic_reference_dataclass():
+def test_cyclic_reference_dataclass(create_module):
     @pydantic.dataclasses.dataclass
     class D1:
         d2: Optional['D2'] = None
 
-    @pydantic.dataclasses.dataclass
-    class D2:
-        d1: Optional[D1] = None
+    @create_module
+    def module():
+        from typing import Optional
 
-    # rebuild_pydantic_dataclass(D1)
+        import pydantic
 
+        @pydantic.dataclasses.dataclass
+        class D2:
+            d1: Optional['D1'] = None
+
+    # Ensure D2 is in the local namespace; note everything works even though it wasn't _defined_ in this namespace
+    D2 = module.D2
+
+    # Confirm D1 and D2 require rebuilding
+    assert isinstance(D1.__pydantic_validator__, MockValidator)
+    assert isinstance(D2.__pydantic_validator__, MockValidator)
+
+    # Note: the rebuilds of D1 and D2 happen automatically, and works since it grabs the locals here as the namespace,
+    # which contains D1 and D2
     instance = D1(d2=D2(d1=D1(d2=D2(d1=D1()))))
 
-    assert TypeAdapter(D1).dump_python(instance) == {...}
+    # Confirm D1 and D2 have been rebuilt
+    assert isinstance(D1.__pydantic_validator__, SchemaValidator)
+    assert isinstance(D2.__pydantic_validator__, SchemaValidator)
+
+    assert TypeAdapter(D1).dump_python(instance) == {'d2': {'d1': {'d2': {'d1': {'d2': None}}}}}
 
     with pytest.raises(ValidationError) as exc_info:
-        D1(d2=D2(d1=D1(d2=D2(d1=D2()))))
-    assert exc_info.value.errors(include_url=False) == [...]
+        D2(d1=D2())
```

Comment:
I removed the parts of the original expression here that don't get hit because the error gets raised while evaluating the inner components of the expression.

Note to self: Before merging, should add some checks that are similar in intent to the original expression, but with a deeply nested dict, to make sure the full recursion is working properly.